crates/ruff/src/rules/flake8_simplify/rules/ast_if.rs

```diff
@@ -627,75 +649,63 @@ pub(crate) fn use_ternary_operator(checker: &mut Checker, stmt: &Stmt, parent: O
     checker.diagnostics.push(diagnostic);
 }
 
-fn get_if_body_pairs<'a>(
-    test: &'a Expr,
-    body: &'a [Stmt],
-    orelse: &'a [Stmt],
-) -> Vec<(&'a Expr, &'a [Stmt])> {
-    let mut pairs = vec![(test, body)];
-    let mut orelse = orelse;
-    loop {
-        if orelse.len() != 1 {
-            break;
-        }
-        let Stmt::If(ast::StmtIf { test, body, orelse: orelse_orelse, range: _ }) = &orelse[0] else {
-            break;
-        };
-        pairs.push((test, body));
-        orelse = orelse_orelse;
-    }
-    pairs
-}
-
 /// SIM114
-pub(crate) fn if_with_same_arms(checker: &mut Checker, stmt: &Stmt, parent: Option<&Stmt>) {
-    let Stmt::If(ast::StmtIf { test, body, orelse, range: _ }) = stmt else {
+pub(crate) fn if_with_same_arms(checker: &mut Checker, locator: &Locator, stmt: &Stmt) {
+    let Stmt::If(ast::StmtIf { body, elif_else_clauses, .. }) = stmt else {
         return;
     };
 
-    // It's part of a bigger if-elif block:
-    // https://github.com/MartinThoma/flake8-simplify/issues/115
-    if let Some(Stmt::If(ast::StmtIf {
-        orelse: parent_orelse,
-        ..
-    })) = parent
-    {
-        if parent_orelse.len() == 1 && stmt == &parent_orelse[0] {
-            // TODO(charlie): These two cases have the same AST:
-            //
-            // if True:
-            //     pass
-            // elif a:
-            //     b = 1
-            // else:
-            //     b = 2
-            //
-            // if True:
-            //     pass
-            // else:
-            //     if a:
-            //         b = 1
-            //     else:
-            //         b = 2
-            //
-            // We want to flag the latter, but not the former. Right now, we flag neither.
-            return;
+    // Check adjacent `if` or `elif` branches, ignore `else`
+    let mut branches_iter = iter::once((stmt.start(), body.as_slice()))
+        .chain(
+            elif_else_clauses
+                .iter()
+                .filter(|clause| clause.test.is_some())
+                .map(|clause| (clause.start(), clause.body.as_slice())),
+        )
+        .peekable();
```

Comment:
Using `peekable` here is probably more expensive (next takes the value from peek in every iteration, just so that we can write it in the immediately followed peek) than keeping track of the last branch in a stack variable

```
let mut Some(last) = branches_iter.next() else {
    return;
}

for (current_start, current_body) in branches_iter {
    ...
    
    last = (current_start, current_body);
}
```

Reply:
don't we lose the first value in this model?

Reply:
but that would compile down to the same thing as the peekable, wouldn't it?

Reply:
That depends on whether the compiler can infer that the `peeked` field on the peekable iterator is always initialized when calling `Iterator` and, thus, can simplify 

```
   #[inline]
    fn next(&mut self) -> Option<I::Item> {
        match self.peeked.take() {
            Some(v) => v,
            None => self.iter.next(),
        }
    }
    ```

to a simple `self.peeked.take()`. It might, it might not. Anwyay, I should have marked this as a Nit. Keep it as you prefer, I don't expect it to change performance significantly (it's not a hot loop, the eliminated code is minimal)